### PR TITLE
[design] Apache Cassandra Design Doc

### LINKED ISSUE #1910: Apache Cassandra Operator
**Is this a bug report or feature request?**
* Feature Request
Use the rook framework and leverage the existing codebase, in order to integrate Cassandra with Kubernetes. 
Cassandra is a highly availably, P2P, distributed database. While it is incredibly performant, running a Cassandra cluster is a painfully manual job. We will take the operational knowledge about running Cassandra and automate it with Kubernetes using the rook framework. 

**What should the feature do:**
A fully functional operator should be able to do the following:
1. Cluster Creation
2. Cassandra Node Replacement
3. (Auto)Scaling
4. Schedule node repairs
5. Backup/Restore
6. Provide a metrics endpoint for Prometheus 

**What is use case behind this feature:**
For rook users, this feature means they get access to one of the most popular NoSQL databases, coupled with a user-friendly UI to manage their self-healing cluster. 
For the rook project, this feature will provide a guide on how to deeply integrate NoSQL databases with Kubernetes, using the rook framework.

### PR DESCRIPTION
**Pull request by yanniszark**
**Description of your changes:**
Added the general design document for Apache Cassandra to the `design/cassandra` folder. All Cassandra related design docs will go to this folder. Any media included in those docs will go to `design/cassandra/media`.  

This document includes the general design without too much detail on any one part. Each stage of development will be accompanied by a doc pertaining to the design choices made. 

**Which issue is resolved by this Pull Request:**
Resolves #1910 

**Checklist:**
- [x] Documentation has been updated, if necessary.

### CODE REVIEW

File: design/cassandra/design.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        requests:
+          cpu: "2000m"
+          memory: "4Gi"
+        limits:
+          cpu: "2000m"
+          memory: "4Gi"
+  backup:
+    storageType: "S3"
+    schedule: "*/30 * * * *"
+  image:
```
Comment (rohan47) [inline design/cassandra/design.md]: Is it necessary to specify the images here?
This can be taken care of by the Cassandra operator code.

File: design/cassandra/design.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+* Cassandra abstracts resources in the following order:
+`Cluster` -> `Datacenter` -> `Rack` -> `Node`
+
+* We map those abstractions to Kubernetes in the following way:
+
+|  Cassandra    |  Kubernetes   |
+|   ----        |     ----      |
+|    Cluster    |  Cluster CRD  |
+|   Datacenter  |  StatefulSet  |
```
Comment (jbw976) [inline design/cassandra/design.md]: It looks like we are trying to tackle the multi-cluster (federation) problem with this design where we are not just managing the cassandra deployment within just 1 k8s cluster, but across a set of clusters, but please correct me if I'm wrong.  There are some k8s community efforts towards that, but that problem space is still in its infancy.  See https://github.com/kubernetes-sigs/federation-v2 for some of the recent thinking.

Do you see value in managing Cassandra within a single k8s cluster (I personally suspect there is definite value)?  That may be where we need to start for the initial implementation until multi-cluster management makes more headway.

File: design/cassandra/design.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+### Sidecar
+
+In Cassandra, many actions require access to the cassandra process or the underlying system. More specifically:
+
+* *Cluster Creation:* dynamically insert java classes and other preparations for Cassandra to run. 
+* *Cassandra Node Replacement:* requires editing config files.
+* *Backup, Monitoring:* requires talking to a JMX interface that the Cassandra process exposes.
+
+By running a sidecar alongside Cassandra, we can monitor the process closely and deal with complex failure scenarios.
```
Comment (jbw976) [inline design/cassandra/design.md]: Can we get into some more details about the sidecar? What binaries/tools need to be in the sidecar container image?  How does it know when to do work (does it watch k8s API, does it expose a REST endpoint of its own, etc.)? How does it talk to the Cassandra node instance it's running along side, is that through exec'ing to a Cassandra standard tool, or does it talk a wire protocol, etc.?

File: design/cassandra/design.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+apiVersion: "cassandra.rook.io/v1alpha1"
+kind: "Cluster"
+metadata:
+  name: "my-cassandra-cluster"
+spec:
+  version: "0.1"
+  dataCenters:
+  - name: "dc-1"
+    racks:
+    - name: "rack-1"
```
Comment (jbw976) [inline design/cassandra/design.md]: I definitely see opportunity to either use, or further refine, the commonality that is captured in the StorageScopeSpec from `rook.io/v1alpha2`: https://github.com/rook/rook/blob/release-0.8/pkg/apis/rook.io/v1alpha2/types.go#L31

The intent of that type is to capture the "boundaries" of the storage substrate, and there is value in using a normalized/common implementation across providers.  Let's see if we can continue this effort of storage resource normalization :)

File: design/cassandra/design.md (hunk 1 above)
Comment (yanniszark) [reply design/cassandra/design.md]: I think it is useful to expose the Cassandra image version, so that the user can have the freedom of choice. My goal is for the operator to use the [official](https://hub.docker.com/_/cassandra/) Cassandra images with a custom entrypoint.
I would agree that we could make this optional and resort to a reasonable default (eg the latest tag for each image).

File: design/cassandra/design.md (hunk 2 above)
Comment (yanniszark) [reply design/cassandra/design.md]: I am aware of that fact and right now, the most you can do is run across availability zones. 
However, even managing Cassandra within a single Kubernetes cluster has very high value, as the database is very well suited for write-heavy workloads. 
If we get to the point where we have a self-healing Cassandra Datacenter, then we can look into configuring multiple Cassandra DCs in different K8s clusters without requiring Federation, as the management of each Cassandra DC is mostly independent of what other DCs are doing.

So, to sum up all of the above, the initial end goal is a single self-healing Cassandra DC running in a single Kubernetes Cluster. Multiple Cassandra Racks (usually translated to Availability Zones) should also be supported as a single K8s cluster can span Availability Zones.

File: design/cassandra/design.md (hunk 3 above)
Comment (yanniszark) [reply design/cassandra/design.md]: After much thought, I think a good approach would be for each sidecar to be a mini-controller responsible for its Cassandra instance. Each sidecar will have its own (Spec, Status), where Spec will be updated by the central controller and Status will be updated by the sidecar itself. That means the sidecar will include client-go and will watch the Kubernetes API. Since it will only watch its own CRD, I expect it to be lightweight. The communication with Cassandra will be done through a socket (Unix preferably, else TCP) using a JMX-HTTP bridge ([Jolokia](https://jolokia.org/)).

File: design/cassandra/design.md [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+The operator will provide all the important features of Cassandra while exposing a user-friendly declarative UI, in accordance to Kubernetes principles.
+More specifically, those features include:
+
+1. Cluster Creation
+2. (Auto)Scaling 
+3. Failed Cassandra Node Replacement
+4. Schedule Node Repair
+5. Backup/Restore
+6. Metrics endpoint for Kubernetes
+
```
Comment (bassam) [inline design/cassandra/design.md]: Are there other Cassandra operators out there that you can list here?

File: design/cassandra/design.md (hunk 2 above)
Comment (bassam) [reply design/cassandra/design.md]: I think it make sense to break the problem into two parts -- managing Cassandra inside a single Kubernetes cluster, and managing Cassandra across Kubernetes cluster. I think we should start with the former.

File: design/cassandra/design.md [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+``` yaml
+apiVersion: "cassandra.rook.io/v1alpha1"
+kind: "Cluster"
+metadata:
+  name: "my-cassandra-cluster"
+spec:
+  # Version will be used to pull the corresponding image from the official Cassandra DockerHub account
+  version: "3.1.11"
+  dataCenters:
```
Comment (bassam) [inline design/cassandra/design.md]: which parts of the spec here are shared with other rook storage providers?

File: design/cassandra/design.md (hunk 3 above)
Comment (bassam) [reply design/cassandra/design.md]: Are you using sidecars as a control channel so that the central Rook operator can communicate with each instance of Cassandra running in the cluster?

File: design/cassandra/sidecar.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+### Examples
+
+Let's consider the case of creating a new Cassandra Cluster. It will look something like this:
+
+1. *User* creates CRD for a Cassandra Cluster.
+2. *Controller* sees the newly created CRD object and creates a StatefulSet for each Cassandra Rack and a clusterIP service for each Pod. Also creates the objects for the sidecars, which will look something like this:
+```yaml
+apiVersion: "cassandra.rook.io/v1alpha1"
+kind: "Sidecar"
```
Comment (bassam) [inline design/cassandra/sidecar.md]: I think the intention here is that this is a `CassandraInstance` or `CassandraDaemon`. I think `Sidecar` is how its implemented not what the Resource is.

File: design/cassandra/sidecar.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+3. *Cassandra* container starts and our custom entrypoint is entered. It waits for config files to be written to a predefined location (shared volume - emptyDir), then copies them to the correct location and starts. 
+
+4. *Sidecar* starts, syncs with the Kubernetes API and gets its (Spec,Status). Checks if the config files are in the shared location, if they are it starts from 4.
+   1. Change state to "Bootstrapping".
+   2. Generate config files with our custom options and copy them to a predefined location (shared volume - emptyDir).
+   3. Poll status through JMX (Jolokia) until the node finishes bootstrapping and is ready.
+   4. Change state to "Ready".
+
+
```
Comment (bassam) [inline design/cassandra/sidecar.md]: neat design! 😀

File: design/cassandra/sidecar.md (hunk 7 above)
Comment (yanniszark) [reply design/cassandra/sidecar.md]: I have discussed about it with @vkoukis, we are working together as part of my internship in @arrikto. He made a similar comment it would be better to name the resource something like "Member" or "CassandraMember". Since you also made the same remark, I guess it would be more accurate to use a different description. This also brings me to another question:

Is it preferable to name the resource "Cluster" and "Member" and distinguish them from the apiVersion, or is it better to name them "CassandraCluster" and "CassandraMember" ? In the first case, `kubectl get clusters` will not work in a namespace with other cluster objects.

File: design/cassandra/design.md (hunk 3 above)
Comment (yanniszark) [reply design/cassandra/design.md]: Essentially, I am using the Sidecar CRDs as a mean of communication between the central controller and each cassandra instance. This is in line with the Kubernetes level driven design and resembles the way the kubelet works. The kubelet watches etcd for Pods which have nodeName set to its node's name, in order to know which Pods it needs to create.

File: design/cassandra/design.md (hunk 2 above)
Comment (yanniszark) [reply design/cassandra/design.md]: I agree ! I will be changing the dataCenters key to dataCenter. It also won't be a list anymore. While we plan to support multiple dataCenters in the future, keeping it this way in the beginning will result in a better UX.

File: design/cassandra/design.md (hunk 6 above)
Comment (yanniszark) [reply design/cassandra/design.md]: I am actively discussing this with members of the rook community.
There is some differentiation there, because I need Cassandra Operator to consume Persistent Volumes (as CockroachDB does) whereas Ceph operator scans the Nodes for local volumes. However, the PlacementSpec should be reusable in Cassandra operator.

File: design/cassandra/design.md (hunk 5 above)
Comment (yanniszark) [reply design/cassandra/design.md]: The main ones I know of are:

* [cassandra-operator](https://github.com/instaclustr/cassandra-operator) by [Instaclustr](https://www.instaclustr.com/): This is indeed a very promising project, coming from a company that offers managed Cassandra deployments. It also utilizes a sidecar, which accepts commands through HTTP endpoints. It is written in Java, so it misses on the more advanced functionality of client-go. Also I don't know how many people are out there developing Kubernetes operators in other than Golang languages, so that may be a barrier for the community the operator will attract. They also do not support different racks.
  * *Phase:* Alpha
  * *License:* Apache 2.0

* [navigator](https://github.com/jetstack/navigator) by [Jetstack](https://www.jetstack.io/): Navigator is aspiring to be a managed DBaaS platform on top of Kubernetes. Their model they promote is very similar to the one I am proposing (CRD + Controller + Sidecar) and has been an inspiration while designing the model for Cassandra Operator. They currently provide operators for Cassandra and Elasticsearch.
  * *Phase:* Alpha
  * *License:* Apache 2.0

There are also a couple of Helm charts out there that try to run Cassandra using StatefulSets only, utilizing things like lifecycle hooks with custom scripts. While they certainly can work, they lack the amount of control that an operator provides. 

Major Point of Differentiation:
* *Static IP:* While StatefulSets provide a stable identity through an FQDN, this is unfortunately not enough for Cassandra, as it identifies its instances by their IP address. Starting a Pod without providing a thing that will serve as the Pod's unique ID makes you subject to race conditions in edge case scenarios. 
  * *Example:* A Cassandra pod starts and bootstraps, but the node fails before you manage to get the node identity (UUID) from the JMX interface. There is no way to know what that instance's ID was, so you can find its IP from gossip to replace it. You can maybe look for dead instances and replace them, but what if there are multiple of them ? This will require manual intervention. By having a Static IP on each Cassandra Pod, we simplify the process of replacing an instance while avoiding race conditions by design, because *we* provide the instance's identity.

I also believe this operator will provide the rook framework with a new pattern (CRD + Controller + Sidecar) that other storage providers can leverage to build their operators.

Comment (Arqu) [reply design/cassandra/design.md]: For what it's worth right now, I'm working on setting up a Scylla cluster on Kubernetes, though with a more manual approach with stateful sets. The last thing I was working on was resolving various networking issues. You are pretty much spot on in terms of the issues that happen due to ephemeral IPs due to the pod's nature. 
Couple things I currently have set up to work around this. Pod and Host IPs resolve differently and sometimes create issues while gossiping due to the dual IPs. Ie if a pod dies and gets restarted, on join it broadcasts with the host IP instead of pod IP. After a long enough period, all of them have converged to their host IPs. On the next pod restart, it fails to figure out the seeds and cluster info thus starts failing. The resolution is to use host networking so the pod binds to the network stack natively and unifies the address. However, this introduces a bunch of other issues. One of the main ones being that with host networking you lose the DNS resolver from Kubernetes thus can no longer resolve nodes by its hostname which brings us to the issue of changing IPs with no mechanism to fixate them. The real solution to the issue is currently pending a PR (https://github.com/kubernetes/kubernetes/pull/68293) which transfers the responsibility to K8s to manage the DNS for the hostname in regards to the changing IPs.
A temporary workaround is to use static IPs per node and then label the nodes and use a selector to force individual instances (at least the seed nodes) on those known nodes. 
In the meanwhile, host networking is at least a bit more stable as full node swaps and restarts are rare (at least in regard to pods) and thus less prone to the issue but are still susceptible unless you hack around it which almost beats the purpose of using orchestration at all. Anyways, once the PR is merged into a stable release it will be possible to manage the entire network setup through K8s without hardcoded bindings and dependencies.

Feel free to reach out for more info as I intend to stick around thanks to the awesome guys at Scylla.

File: design/cassandra/design.md [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+* The operator will create a StatefulSet for each (Datacenter, Rack) tupple.
+* In practice, Datacenter is usually mapped to a Region and Rack to an Availability Zone.
+
+### Major Pain Point: Stable Pod Identity
+
+A problem that every approach so far has, is dealing with loss of persistence.
+If a Cassandra instance loses its data (ie the underlying node fails, the disk dies).
+The replacing instance has to provide the IP of the instance it's replacing. Cassandra doesn't support stable hostnames as an identification method.
+However, IPs on Kubernetes are ephemeral and that leaves us with 2 options:
+
```
Comment (Arqu) [inline design/cassandra/design.md]: Part of this problem can be solved by using persistent volume claims and mapping to stable names/labels provided by the stateful sets. Obviously, this just shifts the blame around, however, in the general case this allows to proverbially save the data which leaves us with only having to map IP addresses. Ie replace them which brings us to your bookkeeping point.
In any case, having the option of PVC vs Local Discs is a sensible solution that can be left as a setting to the end user IMHO but for now, we can avoid feature creep till we have something stable.

File: design/cassandra/design.md [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+* **Sticky IPs Workaround / Service per Instance:** we create a `ClusterIP` Service for each Cassandra instance(Pod). This essentially provides us with a stable IP address for each instance. Since Cassandra uses IP addresses as the means of identification, this means that we don't need to do any bookkeeping. Also, we are immune from race-conditions, as we know our instance's identity beforehand. Some sceptical thoughts about this method:
+  1. Feels/Seems hacky
+  2. Uses too many services: most clusters have a /12 IP block to utilize, so we do not expect them to run out of IPs soon.
+  3. Performance Issues: since the default implementation of ClusterIPs is iptables this could be an issue with clusters in the 100s/1000s. However, an IPVS implementation of ClusterIP is [GA in 1.11](https://kubernetes.io/blog/2018/07/09/ipvs-based-in-cluster-load-balancing-deep-dive/) and works well even with 1000s of Services. 
+
```
Comment (Arqu) [inline design/cassandra/design.md]: Currently I'm experiencing some issues running this setup (manually though) where the ClusterIP is not the same as the Node/PodID thus resulting in some confusion regarding the resolution of nodes both internally as well as to outside services. I think that this could potentially be worked around if we could bind the service address to the nodes, however, I don't think that is possible as of now. I'd be cautious with supporting this in the longer term and would need extensive testing to be trustworthy and that is beside the named caveats.

On the flip side, I still prefer this approach to the "manual" bookkeeping as I think that will bring a host of unforeseen issues.

Also just for reference sake, this would be a blessing for our use case:
https://github.com/kubernetes/kubernetes/issues/28969

File: design/cassandra/sidecar.md [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+      * Doesn't support nested fields
+  2. **Member CRD**
+    * Each sidecar will watch an instance of a newly defined Member CRD and have its own `Spec` and `Status`.
+    * **Advantages:**
+      * More expressive and natural. Supports nested fields.
+      * Only our operator touches it. We don't expect it to happen often, but Pods are touched by pretty much everyone on the cluster. So if someone does stupid things, that affects us too.
+    * **Disadvantages:**
+      * Probably overkill to have for only a couple of fields.
+      * Induces an extra burden on etcd.
+
```
Comment (Arqu) [inline design/cassandra/sidecar.md]: I would prefer the first approach here, keeping it simple will probably make a lot of future things much simpler as well. Opening the config up for external messing seems like a great way for extra issues to be opened up. Also, nested fields don't really seem like a pressing requirement as far as I can see this. Worst case if we do happen to need such functionality is to use combinations of the fields.

With labels it pretty much a state machine which would be very easy to map out in our case in terms of flow of actions.

File: design/cassandra/design.md (hunk 10 above)
Comment (yanniszark) [reply design/cassandra/design.md]: Thanks for your comments @Arqu !
I'll try answering some of your concerns:
1. If you provide the ClusterIP as the `broadcast_address` and the Pod IP as the `listen_address` it shouldn't exhibit any unpredictable behaviour. I have a manual setup like this up and working.
2. As for exposing things to the outside world, this will most likely be done though an Ingress. Generally, exposing Cassandra to services outside Kubernetes is an issue for every approach, so we should deal with this when the time comes. A simple workaround I can think of is to use a Cassandra driver that gets the data from the coordinator node and doesn't try to find the owner of the range.

**Comment by jbw976** [comment]
@yanniszark are there updates that should be made to this design doc to bring it into agreement with the implementation in #2207?

Should this design be consolidated with the design in #2135?  Sorry for the delay in reviewing this Cassandra work, I'll be focusing on it now to keep up all your recent momentum :)

**Comment by yanniszark** [comment]
> @yanniszark are there updates that should be made to this design doc to bring it into agreement with the implementation in #2207?
> 
> Should this design be consolidated with the design in #2135? Sorry for the delay in reviewing this Cassandra work, I'll be focusing on it now to keep up all your recent momentum :)

@jbw976 everything is up-to-date with the implementation. 
The design in #2135 is a detailed design of the Cluster Creation sequence. This is a more general design document about the overall architecture. I was thinking that each phase of development (eg. cluster creation, scaling, monitoring etc.) should have its own design doc. All of those docs will be under the `design/cassandra` directory.

File: design/cassandra/sidecar.md [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+* **Answer:** No, it writes the `nodeName` field on the PodSpec. In other words, it writes a record of intent. No matter how many times the kubelet or the scheduler crashes it doesn't matter. The record of intent is there. 
+
+### Control-Loop Design
+
+* Based on our observations above, we design a method of communication in line with the Kubernetes philosophy.
+* When the controller wants to invoke a functionality in a Sidecar, it should write a record of intent in the Kubernetes API Objects (etcd). The sidecar will be watching the Kubernetes API and responding accordingly.
+* There are two approaches to represent the record of intent:
+  1. **Labels:**
+    * When the controller wants to communicate with a sidecar, it will write a predefined label in the ClusterIP Service Object of the specific instance. For example, to communicate that we want an instance to terminate, we could write the label 'cassandra.rook.io/termination-timestamp=0`. The sidecar will see this and decommission the Cassandra instance. When it is done, it will change the label value to a valid timestamp. Then the controller will know to delete that instance.
```
[the remark this answers is not in the document]
Comment (yanniszark) [reply design/cassandra/sidecar.md]: That's not really defined somewhere atm.
I will add documentation as comments in the `pkg/operator/cassandra/constants` package.
It may also be useful to document them in a separate file.

File: design/cassandra/design.md [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+* Cassandra abstracts resources in the following order:
+`Cluster` -> `Datacenter` -> `Rack` -> `Node`
+
+* We map those abstractions to Kubernetes in the following way:
+
+|  Cassandra    |  Kubernetes   |
+|   ----        |     ----      |
+|    Cluster    |  Cluster CRD  |
+|   Datacenter  |  StatefulSet  |
```
[the remark this answers is not in the document]
Comment (yanniszark) [reply design/cassandra/design.md]: You are correct, this is a mistake in the docs. 
Datacenter is a group of StatefulSets.
Rack is a single StatefulSet.

File: design/cassandra/sidecar.md [hunk 14]
```diff
@@ -0,0 +1,79 @@
+## Sidecar Design Proposal
+
+### Consideration: REST API
+
+When thinking about how our sidecar will communicate with our controller, a natural solution that comes to mind is though a REST API. The sidecar will run an HTTP Server which the other party will call. This is the approach used by [Netflix's Priam](https://github.com/Netflix/priam/wiki).
```
[the remark this answers is not in the document]
Comment (yanniszark) [reply design/cassandra/sidecar.md]: No, it doesn't make HTTP calls to the sidecar at all. 

Cassandra exposes a JMX interface. We can't talk JMX, so we apply a JMX-HTTP proxy (jolokia) and talk through HTTP. Inside the Pod, Cassandra serves an HTTP endpoint on localhost and the sidecar consumes that in order to talk to Cassandra. I'd love some advice on how to make this more clear from the code/docs.

**Comment by jbw976** [comment]
I'm thinking it's time to merge this and #2135.  Then we can probably go ahead and close #1910.  What do you think @yanniszark?

**Comment by yanniszark** [comment]
@jbw976 I agree that we can merge this.
I have fixed a few minor things and I think it's ready merging.
